lib/spack/llnl/util/lang.py

```diff
@@ -166,30 +166,24 @@ def union_dicts(*dicts):
     return result
 
 
-class memoized(object):
-    """Decorator that caches the results of a function, storing them
-       in an attribute of that function."""
-
-    def __init__(self, func):
-        self.func = func
-        self.cache = {}
+def memoized(func):
```

Comment:
Without 3960c1d doc tests will fail as Sphinx is not able to extract the signature of the methods that have been decorated. Implementing it as a function solves the issue.

Reply:
From what I can tell, all uses of `memoized` would have had this problem, so it's interesting that the doctests fail now, do you know what the difference is for the newly-decorated functions?

I can see how docstrings would have all been omitted without the use of `functools.wraps` (or possibly with `functools.update_wrapper`.

Also, this seems like it would be good to maintain as a separate commit, vs. squashing; on the other hand there's at least a couple commits (like https://github.com/spack/spack/pull/10931/commits/8948d3db30aaf803277ff4ccdf3f034b66e6f2c9) that should be squashed. Could you rebase this?

Reply:
> From what I can tell, all uses of memoized would have had this problem, so it's interesting that the doctests fail now, do you know what the difference is for the newly-decorated functions?

I didn't dig deep into the error, as it seemed worthwhile to add `functools.wrap` anyhow. I think this is the only place where `memoized` is used with a `@classmethod` and my guess is that there's something that has to do with both decorators applied to a method.

> Also, this seems like it would be good to maintain as a separate commit, vs. squashing; on the other hand there's at least a couple commits (like 8948d3d) that should be squashed. Could you rebase this?

I'll extract this into its own PR then rebase, if necessary, this one as soon as the other modification gets merged.

Reply:
Should be ready